File: bank_retenue_sync/bank_retenue_sync/report/carte_technologique/carte_technologique.py
"""Suivi de la carte technologique : ce qui la recharge, ce qui la depense, ce qui y reste.

POURQUOI UN SUIVI A PART
------------------------
La carte technologique est une TRESORERIE PARALLELE : elle est alimentee par un virement depuis
Zitouna (« CHARGEMENT CARTE TECHNOLOGIQUE » au releve) puis depensee en publicite Facebook, et ces
depenses-la ne paraissent JAMAIS au releve bancaire. Le rapprochement bancaire ne peut donc rien
en dire — d'ou ce suivi, qui lit le compte lui-meme.

Un solde de carte qui approche zero est une information operationnelle, pas comptable : c'est le
moment de la recharger avant que les campagnes ne s'arretent. C'est ce que la tuile « Solde
disponible » donne d'un coup d'oeil.

LES DEUX SENS N'ONT PAS LA MEME PREUVE
--------------------------------------
Une RECHARGE se prouve au releve : elle a une reference bancaire, et `bank/classify` la rapproche.
Une DEPENSE n'a que sa facture — rien en banque ne la confirme. La colonne « Rapproché » ne
concerne donc que les recharges ; l'afficher sur une depense laisserait croire a un controle qui
n'existe pas.
"""
from __future__ import annotations

import frappe
from frappe import _
from frappe.utils import flt, getdate

COMPTE_CARTE = "Carte technologique - A&S"


def execute(filters=None):
    filters = frappe._dict(filters or {})
    compte = filters.get("compte") or COMPTE_CARTE
    lignes = _mouvements(compte, filters)
    return _colonnes(), lignes, None, _graphique(lignes), _synthese(compte, lignes)


def _colonnes() -> list:
    return [
        {"fieldname": "date", "label": _("Date"), "fieldtype": "Date", "width": 100},
        {"fieldname": "nature", "label": _("Nature"), "fieldtype": "Data", "width": 110},
        {"fieldname": "libelle", "label": _("Libellé"), "fieldtype": "Data", "width": 280},
        {"fieldname": "recharge", "label": _("Recharge"), "fieldtype": "Currency", "width": 120},
        {"fieldname": "depense", "label": _("Dépense"), "fieldtype": "Currency", "width": 120},
        {"fieldname": "solde", "label": _("Solde après"), "fieldtype": "Currency", "width": 130},
        {"fieldname": "journal_entry", "label": _("Écriture"), "fieldtype": "Link",
         "options": "Journal Entry", "width": 170},
        {"fieldname": "rapproche", "label": _("Rapproché"), "fieldtype": "Data", "width": 110},
    ]


def _mouvements(compte: str, filters) -> list:
    conditions, params = ["gl.account = %(compte)s", "gl.is_cancelled = 0"], {"compte": compte}
    if filters.get("date_from"):
        conditions.append("gl.posting_date >= %(date_from)s")
        params["date_from"] = filters.date_from
    if filters.get("date_to"):
        conditions.append("gl.posting_date <= %(date_to)s")
        params["date_to"] = filters.date_to

    rows = frappe.db.sql(
        """select gl.posting_date, gl.voucher_type, gl.voucher_no,
                  round(sum(gl.debit), 3) as recharge, round(sum(gl.credit), 3) as depense
           from `tabGL Entry` gl
           where %s
           group by gl.voucher_type, gl.voucher_no, gl.posting_date
           order by gl.posting_date, gl.voucher_no""" % " and ".join(conditions),
        params, as_dict=1)
    if not rows:
        return []

    textes = _textes([r.voucher_no for r in rows])
    # Une recharge est rapprochee si un mouvement du releve pointe son ecriture. C'est le seul
    # controle possible ici : les depenses n'ont pas de contrepartie bancaire.
    rapproches = set(frappe.db.get_all(
        "BRS Bank Movement", filters={"document_name": ["in", [r.voucher_no for r in rows]]},
        pluck="document_name"))
    # ⚠️ Hors de la periode du registre, l'absence de rapprochement ne prouve RIEN : le releve
    # ne remonte pas jusque-la. Sans cette borne, 18 recharges de 2024-2025 ressortaient en
    # alerte rouge alors qu'elles sont simplement anterieures a l'import.
    debut, fin = frappe.db.sql(
        "select min(`date`), max(`date`) from `tabBRS Bank Movement`")[0]

    # Le solde progressif est calcule sur TOUTE l'histoire du compte, meme si un filtre de date
    # masque le debut : un solde qui repartirait de zero au 1er du mois serait faux.
    solde = _solde_avant(compte, rows[0].posting_date)
    out = []
    for r in rows:
        solde = round(solde + r.recharge - r.depense, 3)
        recharge = r.recharge > 0
        out.append({
            "date": r.posting_date,
            "nature": _("Recharge") if recharge else _("Dépense"),
            "libelle": textes.get(r.voucher_no) or r.voucher_no,
            "recharge": r.recharge or None,
            "depense": r.depense or None,
            "solde": solde,
            "journal_entry": r.voucher_no if r.voucher_type == "Journal Entry" else None,
            "rapproche": _rapproche(r, recharge, rapproches, debut, fin),
        })
    return out


def _rapproche(r, recharge: bool, rapproches: set, debut, fin) -> str:
    """« oui » / « non » / « » — et le vide n'est pas un oubli, c'est l'absence de preuve."""
    if not recharge:
        return ""
    if r.voucher_no in rapproches:
        return _("oui")
    if not debut or not (getdate(debut) <= getdate(r.posting_date) <= getdate(fin)):
        return _("hors registre")
    return _("non")


def _solde_avant(compte: str, jour) -> float:
    return flt(frappe.db.sql(
        """select sum(debit) - sum(credit) from `tabGL Entry`
           where account = %s and is_cancelled = 0 and posting_date < %s""",
        (compte, getdate(jour)))[0][0], 3)


def _textes(noms: list) -> dict:
    return {r.name: (r.cheque_no or r.user_remark or "").split("\n")[0][:140]
            for r in frappe.db.get_all("Journal Entry", filters={"name": ["in", noms]},
                                       fields=["name", "cheque_no", "user_remark"],
                                       limit_page_length=0)}


def _synthese(compte: str, lignes: list) -> list:
    """Le solde vient du COMPTE, pas des lignes affichees : un filtre de date ne doit pas
    changer l'argent disponible sur la carte."""
    total = frappe.db.sql(
        """select round(sum(debit), 3), round(sum(credit), 3), round(sum(debit) - sum(credit), 3)
           from `tabGL Entry` where account = %s and is_cancelled = 0""", compte)[0]
    recharges, depenses, solde = (flt(x, 3) for x in total)
    non_rapprochees = [l for l in lignes if l["rapproche"] == _("non")]
    tuiles = [
        {"value": solde, "label": _("Solde comptable"), "datatype": "Currency",
         "indicator": "Red" if solde < 200 else ("Orange" if solde < 500 else "Green")},
    ]
    # Le solde REEL vient du service : le confronter au solde comptable est le seul controle
    # possible sur ce compte, la carte n'ayant aucune contrepartie au releve bancaire. C'est lui
    # qui a revele deux paiements de juillet jamais comptabilises.
    tuiles += _tuiles_controle(solde)
    return tuiles + [
        {"value": flt(sum(l["depense"] or 0 for l in lignes), 3),
         "label": _("Dépensé sur la période"), "datatype": "Currency"},
        {"value": flt(sum(l["recharge"] or 0 for l in lignes), 3),
         "label": _("Rechargé sur la période"), "datatype": "Currency"},
        {"value": len(non_rapprochees), "label": _("Recharges non rapprochées"), "datatype": "Int",
         "indicator": "Red" if non_rapprochees else "Green"},
        {"value": recharges, "label": _("Total rechargé"), "datatype": "Currency"},
        {"value": depenses, "label": _("Total dépensé"), "datatype": "Currency"},
    ]


def _tuiles_controle(solde_comptable: float) -> list:
    """Solde reel, ecart, et surtout : combien de paiements du releve manquent aux livres.

    Non bloquant par contrat : le service peut etre injoignable (autre conteneur, scraping en
    cours). Le rapport doit rester lisible sans lui — on perd le controle, pas le suivi.

    L'ordre des tuiles suit l'ordre du raisonnement : le nombre d'operations manquantes vient
    AVANT l'ecart, parce qu'il l'explique. Un ecart de 620 DT avec un paiement non comptabilise
    n'est pas un frais bancaire a regulariser, c'est ce paiement — les lire dans l'autre sens
    conduit tout droit a inventer une charge et a masquer l'operation.
    """
    from bank_retenue_sync.bank import cartes

    try:
        etat = cartes.etat_controle()
    except Exception:
        return []
    r = etat["resume"]
    tuiles = [
        {"value": r["a_comptabiliser"], "label": _("Paiements carte non comptabilisés"),
         "datatype": "Int", "indicator": "Red" if r["a_comptabiliser"] else "Green"},
    ]
    if r["a_comptabiliser"]:
        tuiles.append({"value": r["montant_a_comptabiliser"], "label": _("Montant à comptabiliser"),
                       "datatype": "Currency", "indicator": "Red"})
    if etat.get("lu_le"):
        reel, ecart = flt(etat["solde_reel"], 3), flt(etat["ecart"], 3)
        tuiles += [
            {"value": reel, "label": _("Solde réel (carte)"), "datatype": "Currency",
             "indicator": "Red" if reel < 200 else "Green"},
            {"value": ecart, "label": _("Écart livres / carte"), "datatype": "Currency",
             "indicator": "Green" if abs(ecart) < 0.5 else "Red"},
            {"value": flt(etat.get("plafond_restant"), 3), "label": _("Plafond annuel restant"),
             "datatype": "Currency"},
        ]
    return tuiles


def _graphique(lignes: list) -> dict:
    """Le solde dans le temps : c'est la courbe qui dit quand recharger."""
    if not lignes:
        return None
    return {
        "data": {"labels": [str(l["date"]) for l in lignes],
                 "datasets": [{"name": _("Solde"), "values": [l["solde"] for l in lignes]}]},
        "type": "line", "colors": ["#2490ef"], "lineOptions": {"regionFill": 1},
    }

File: bank_retenue_sync/bank_retenue_sync/report/carte_technologique/carte_technologique.js
// Suivi de la carte technologique. Aucune borne de date par défaut : le solde disponible est la
// question du jour, et il se lit sur toute l'histoire du compte.
//
// Deux actions viennent s'ajouter au tableau, parce que le tableau seul ne répond pas à la
// question qu'on se pose devant lui :
//   « Contrôle du relevé »   — mes livres sont-ils à jour avec la carte ? (lecture seule)
//   « Générer les écritures » — créer celles qui manquent, avec un essai à blanc d'abord.
// Le tableau lit le GRAND LIVRE ; le contrôle lit le RELEVÉ DE CARTE. Ce que le premier ne peut
// pas montrer, par construction, c'est ce qui n'a jamais été saisi — d'où le second.
frappe.query_reports["Carte technologique"] = {
  filters: [
    {
      fieldname: "compte",
      label: __("Compte"),
      fieldtype: "Link",
      options: "Account",
      default: "Carte technologique - A&S",
      reqd: 1,
    },
    { fieldname: "date_from", label: __("Du"), fieldtype: "Date" },
    { fieldname: "date_to", label: __("Au"), fieldtype: "Date" },
  ],

  onload(report) {
    report.page.add_inner_button(__("Contrôle du relevé"), () => controle(report, 0));
    // Rafraîchir déclenche un nouveau scraping chez la banque : c'est long et ça sollicite le
    // service, donc jamais implicite — le contrôle courant lit le dernier export disponible.
    report.page.add_inner_button(__("Contrôle (relevé rafraîchi)"), () => controle(report, 1));
    if (frappe.user.has_role("System Manager")) {
      report.page.add_inner_button(__("Générer les écritures"), () => generer(report));
    }
    // Le verdict « à jour / en retard » est posé sur la page dès l'ouverture : c'est la première
    // chose qu'on veut savoir, et elle ne se déduit d'aucune colonne du tableau.
    etat(0).then((d) => d && indicateur(report, d));
  },

  formatter(value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);
    // Une recharge non rapprochée au relevé est la seule anomalie que ce tableau peut détecter.
    if (data && column.fieldname === "rapproche" && data.rapproche === "non") {
      value = `<span style="color:#a93226;font-weight:600">${value}</span>`;
    }
    // Le solde qui descend est l'information opérationnelle : recharger avant la panne.
    if (data && column.fieldname === "solde" && flt(data.solde) < 500) {
      value = `<span style="color:#b9770e;font-weight:600">${value}</span>`;
    }
    return value;
  },
};

// Lecture seule : ne crée rien, ne modifie rien. Échoue en silence si le service bancaire est
// muet — le suivi comptable doit rester consultable sans lui.
function etat(refresh) {
  return frappe
    .call({
      method: "bank_retenue_sync.bank.cartes.controle_releve",
      args: { refresh: refresh || 0 },
      freeze: !!refresh,
      freeze_message: __("Lecture du relevé de carte chez la banque…"),
    })
    .then((r) => r.message)
    .catch(() => null);
}

function indicateur(report, d) {
  if (d.erreur) {
    report.page.set_indicator(__("Relevé de carte illisible"), "grey");
    return;
  }
  const n = d.resume.a_comptabiliser;
  if (n) {
    report.page.set_indicator(__("{0} paiement(s) non comptabilisé(s)", [n]), "red");
  } else if (d.a_jour) {
    report.page.set_indicator(__("Comptabilité à jour"), "green");
  } else {
    // Rien à saisir, mais les soldes ne concordent pas (ou le service n'a pas répondu) : ni
    // alerte ni feu vert — dire « à jour » ici serait affirmer un contrôle qu'on n'a pas fait.
    report.page.set_indicator(__("Écart à vérifier"), "orange");
  }
}

function controle(report, refresh) {
  etat(refresh).then((d) => {
    if (!d || d.erreur) {
      frappe.msgprint({
        title: __("Contrôle impossible"),
        message:
          __("Le relevé de carte n'a pas pu être lu (service bancaire injoignable).") +
          ((d && d.erreur && `<br><span class="text-muted">${frappe.utils.escape_html(d.erreur)}</span>`) || ""),
        indicator: "orange",
      });
      return;
    }
    indicateur(report, d);
    const dialog = new frappe.ui.Dialog({
      title: __("Contrôle du relevé de carte"),
      size: "extra-large",
      fields: [{ fieldtype: "HTML", fieldname: "corps" }],
    });
    dialog.fields_dict.corps.$wrapper.html(html_controle(d));
    // Le bouton de création n'apparaît que s'il y a quelque chose à créer : proposer « générer »
    // devant un contrôle vert invite à cliquer pour rien.
    if (d.resume.a_comptabiliser && frappe.user.has_role("System Manager")) {
      dialog.set_primary_action(
        __("Générer les {0} écriture(s) manquante(s)", [d.resume.a_comptabiliser]),
        () => {
          dialog.hide();
          generer(report);
        }
      );
    }
    dialog.show();
  });
}

// Le service rend un horodatage ISO (« 2026-08-11T19:12:24 ») que `str_to_user` ne sait pas lire :
// le « T » lui fait rendre une date invalide. La date de lecture du relevé est justement ce qui
// dit si le contrôle porte sur un état récent — l'afficher faux serait pire que ne rien afficher.
function lu_le(d) {
  if (!d.lu_le) return "—";
  return frappe.datetime.str_to_user(String(d.lu_le).replace("T", " ")) || "—";
}

function html_controle(d) {
  const r = d.resume;
  const dt = (v) => format_currency(v, "TND");
  const verdict = r.a_comptabiliser
    ? `<div style="padding:10px;border-radius:4px;background:#fbe9e7;color:#a93226">
         <b>${__("Les livres sont en retard sur la carte.")}</b><br>
         ${__("{0} paiement(s) du relevé, soit {1}, n'ont aucune écriture.", [
           r.a_comptabiliser,
           dt(r.montant_a_comptabiliser),
         ])}
       </div>`
    : d.a_jour
      ? `<div style="padding:10px;border-radius:4px;background:#e8f5e9;color:#1e7e34">
           <b>${__("Comptabilité à jour.")}</b><br>
           ${__("Tous les paiements approuvés du relevé sont comptabilisés, et le solde des livres égale le solde réel de la carte.")}
         </div>`
      : `<div style="padding:10px;border-radius:4px;background:#fff4e5;color:#b9770e">
           <b>${__("Rien à saisir, mais les soldes ne concordent pas.")}</b><br>
           ${__("Aucun paiement du relevé ne manque : l'écart restant est un frais prélevé par la banque, régularisé par la tâche quotidienne.")}
         </div>`;

  const soldes = `
    <table class="table table-bordered" style="margin-top:12px">
      <tr>
        <td>${__("Solde comptable (ERPNext)")}</td><td align="right"><b>${dt(d.solde_comptable)}</b></td>
        <td>${__("Solde réel (carte)")}</td>
        <td align="right"><b>${d.solde_reel === null ? "—" : dt(d.solde_reel)}</b></td>
        <td>${__("Écart")}</td>
        <td align="right"><b style="color:${Math.abs(d.ecart || 0) < 0.5 ? "#1e7e34" : "#a93226"}">
          ${d.ecart === null ? "—" : dt(d.ecart)}</b></td>
      </tr>
      <tr>
        <td>${__("Relevé lu le")}</td><td align="right">${lu_le(d)}</td>
        <td>${__("Plafond annuel restant")}</td>
        <td align="right">${d.plafond_restant === null ? "—" : dt(d.plafond_restant)}</td>
        <td>${__("Seuil de recharge")}</td><td align="right">${dt(d.seuil_recharge)}</td>
      </tr>
    </table>`;

  // Les refus figurent dans la liste et n'en sortiront pas : c'est en les voyant écartés qu'on
  // vérifie qu'ils l'ont été. Sur l'export du 21/07, 8 lignes sur 17 étaient des refus — les
  // comptabiliser aurait inventé 3 438 DT de charges.
  const etats = {
    comptabilisee: [__("comptabilisée"), "#1e7e34"],
    a_comptabiliser: [__("à comptabiliser"), "#a93226"],
    refusee: [__("refusée par la carte"), "#8d8d8d"],
  };
  const lignes = d.lignes
    .map((l) => {
      const [libelle, couleur] = etats[l.etat];
      const je = l.journal_entry
        ? `<a href="/app/journal-entry/${encodeURIComponent(l.journal_entry)}">${l.journal_entry}</a>
           ${l.brouillon ? `<span style="color:#b9770e;font-size:11px"> (${__("brouillon")})</span>` : ""}`
        : "";
      return `<tr style="${l.etat === "refusee" ? "opacity:.6" : ""}">
          <td>${frappe.datetime.str_to_user(l.date)}</td>
          <td>${frappe.utils.escape_html(l.detail || "")}</td>
          <td>${frappe.utils.escape_html(l.statut || "")}</td>
          <td align="right">${dt(l.montant)}</td>
          <td>${frappe.utils.escape_html(l.reference || "")}</td>
          <td style="color:${couleur};font-weight:600">${libelle}</td>
          <td>${je}</td>
        </tr>`;
    })
    .join("");

  return `${verdict}${soldes}
    <table class="table table-bordered" style="font-size:12px">
      <thead><tr>
        <th>${__("Date")}</th><th>${__("Détail")}</th><th>${__("Statut carte")}</th>
        <th align="right">${__("Montant")}</th><th>${__("Référence")}</th>
        <th>${__("État")}</th><th>${__("Écriture")}</th>
      </tr></thead>
      <tbody>${lignes}</tbody>
    </table>
    <div class="text-muted" style="font-size:11px">
      ${__("{0} ligne(s) au relevé : {1} comptabilisée(s), {2} à comptabiliser, {3} refusée(s) par la carte (aucun mouvement d'argent).", [
        r.total,
        r.comptabilisees,
        r.a_comptabiliser,
        r.refusees,
      ])}
    </div>`;
}

// Essai à blanc d'abord, création ensuite. Deux temps volontairement : une écriture de journal
// soumise ne se retire pas d'un clic, et l'utilisateur doit voir CE QUI SERA CRÉÉ avant de le
// créer — le montant, la date et la référence de chaque ligne.
function generer(report) {
  const dialog = new frappe.ui.Dialog({
    title: __("Générer les écritures de la carte"),
    fields: [
      {
        fieldtype: "Check",
        fieldname: "refresh",
        label: __("Rafraîchir le relevé chez la banque avant (plus lent)"),
        default: 0,
        description: __(
          "Sans cette option, le dernier relevé exporté est utilisé. Les paiements du jour même peuvent en être absents."
        ),
      },
      { fieldtype: "HTML", fieldname: "apercu" },
    ],
    primary_action_label: __("Essai à blanc"),
    primary_action: () => {
      const refresh = dialog.get_value("refresh");
      frappe
        .call({
          method: "bank_retenue_sync.orchestrator.run_cartes",
          args: { insert: 0, refresh: refresh ? 1 : 0 },
          freeze: true,
          freeze_message: __("Lecture du relevé et simulation…"),
        })
        .then((r) => {
          const res = r.message || {};
          const a_creer = (res.ecritures || []).filter((e) => e.status === "created");
          dialog.fields_dict.apercu.$wrapper.html(html_apercu(res, a_creer));
          if (!a_creer.length) {
            dialog.set_primary_action(__("Fermer"), () => dialog.hide());
            return;
          }
          // Le relevé a déjà été rafraîchi à l'essai : le refaire à la création rallongerait
          // l'attente et, surtout, ferait porter la création sur un relevé DIFFÉRENT de celui
          // qui vient d'être validé à l'écran.
          dialog.set_primary_action(
            __("Créer les {0} écriture(s)", [a_creer.length]),
            () => {
              frappe
                .call({
                  method: "bank_retenue_sync.orchestrator.run_cartes",
                  args: { insert: 1, refresh: 0 },
                  freeze: true,
                  freeze_message: __("Création des écritures…"),
                })
                .then((r2) => {
                  dialog.hide();
                  const cree = (r2.message.ecritures || []).filter((e) => e.status === "created");
                  frappe.show_alert(
                    { message: __("{0} écriture(s) créée(s)", [cree.length]), indicator: "green" },
                    7
                  );
                  frappe.msgprint({
                    title: __("Écritures créées"),
                    message: html_apercu(r2.message || {}, cree, true),
                    indicator: "green",
                  });
                  report.refresh();
                  etat(0).then((d) => d && indicateur(report, d));
                });
            }
          );
        });
    },
  });
  dialog.fields_dict.apercu.$wrapper.html(
    `<div class="text-muted">${__("Lancez l'essai à blanc pour voir ce qui serait créé. Rien n'est écrit à cette étape.")}</div>`
  );
  dialog.show();
}

function html_apercu(res, ecritures, cree) {
  const dt = (v) => format_currency(v, "TND");
  const erreurs = (res.ecritures || []).filter((e) => e.status === "error");
  if (!ecritures.length && !erreurs.length) {
    return `<div style="padding:10px;border-radius:4px;background:#e8f5e9;color:#1e7e34">
      <b>${__("Aucune écriture à créer.")}</b><br>
      ${__("Tous les paiements approuvés du relevé sont déjà comptabilisés.")}
      ${html_frais(res)}</div>`;
  }
  const lignes = ecritures
    .map(
      (e) => `<tr>
        <td>${frappe.datetime.str_to_user(e.date)}</td>
        <td>${frappe.utils.escape_html(e.ref || "")}</td>
        <td align="right">${dt(e.montant)}</td>
        <td>${cree && e.je !== "(dry-run)" ? `<a href="/app/journal-entry/${encodeURIComponent(e.je)}">${e.je}</a>` : ""}</td>
      </tr>`
    )
    .join("");
  const erreurs_html = erreurs.length
    ? `<div style="color:#a93226;margin-top:8px">${erreurs
        .map((e) => `${frappe.utils.escape_html(e.ref)} : ${frappe.utils.escape_html(e.error)}`)
        .join("<br>")}</div>`
    : "";
  return `
    <div style="margin-bottom:6px">${
      cree
        ? __("Écritures créées — Dr Frais de Marketing / Cr Carte technologique :")
        : __("Seraient créées — Dr Frais de Marketing / Cr Carte technologique :")
    }</div>
    <table class="table table-bordered" style="font-size:12px">
      <thead><tr><th>${__("Date")}</th><th>${__("Référence")}</th>
        <th align="right">${__("Montant")}</th><th>${__("Écriture")}</th></tr></thead>
      <tbody>${lignes}</tbody>
    </table>${erreurs_html}${html_frais(res)}`;
}

// La régularisation de frais n'est PAS un paiement : elle aligne le solde comptable sur le solde
// réel, et seulement une fois toutes les opérations reconnues. La montrer à part évite de la
// confondre avec une charge de publicité.
function html_frais(res) {
  const f = res.frais;
  if (!f || !f.statut) return "";
  const messages = {
    aligne: __("Soldes alignés : aucun frais à régulariser."),
    "operations non comptabilisees": __(
      "Frais non régularisés : des paiements du relevé restent à comptabiliser. L'écart n'est pas un frais tant qu'ils manquent."
    ),
    "deja regularise": __("Frais déjà régularisés pour cette lecture du solde."),
    "a creer": __("Un frais bancaire de {0} serait créé pour aligner les soldes.", [
      format_currency(f.montant, "TND"),
    ]),
    regularise: __("Frais bancaire de {0} créé ({1}).", [format_currency(f.montant, "TND"), f.je]),
  };
  const m = messages[f.statut] || f.statut;
  return `<div class="text-muted" style="font-size:11px;margin-top:8px">${m}</div>`;
}
